src/main/java/org/assertj/core/api/AssertFactory.java

```diff
@@ -12,28 +12,21 @@
  */
 package org.assertj.core.api;
 
+import java.util.function.Function;
+
 /**
- * A single method factory interface to create an {@link Assert} class for a given type.
- * This factory method typically wraps a call to <code>assertThat(t)</code> to map to the concrete assert type {@code ASSERT}
- * for the element {@code T}.
+ * A {@link Function} to produce an {@link Assert} for a given value.
+ * This function typically wraps a call to <code>assertThat(value)</code> to produce a concrete assert type {@code ASSERT}
+ * for the input value.
  * <p>
- * This interface is typically used by navigation assertions on iterable types like {@link AbstractIterableAssert} when calling
+ * This function is typically used by navigation assertions on iterable types like {@link AbstractIterableAssert} when calling
  * {@link Assertions#assertThat(Iterable, AssertFactory) assertThat(Iterable&lt;E&gt;, AssertFactory&lt;E, ASSERT&gt;)}
  * <p>
- * @param <T> the type to create assertions for.
- * @param <ASSERT> the assertions class for T.
+ * @param <T> the type of the input to the function.
+ * @param <ASSERT> the type of the resulting {@code Assert}.
  * 
  * @since 2.5.0 / 3.5.0
  */
 @FunctionalInterface
-public interface AssertFactory<T, ASSERT> {
-
-  /**
-   * Creates the custom Assert object for the given element value.
-   *
-   * Typically this will just invoke <code>assertThat(t)</code>
-   * @param t the type to convert to an Assert object
-   * @return returns <code>assertThat(t)</code>
-   */
-  ASSERT createAssert(T t);
+public interface AssertFactory<T, ASSERT extends Assert<?, ?>> extends Function<T, ASSERT> {
```

Comment:
My 2 cents about this change. Doing this change is a backwards incompatible change:

* Removal of the `createAssert` method
* Changing the generic `ASSERT` to extend `Assert`

IMHO changes to this interface is not really needed.

In theory the `AssertFactory` itself could be annotated with `@FunctionalInterface`

Reply:
IMHO without those changes the semantic of this interface was wrong, there was nothing ensuring that the produced object was an instance an `Assert`. The interface was basically only a copy of `Function` with different method name.

With these changes the assert concept is enforced and `AssertFactory` gets any existing default method in `Function`. (BTW, it was already a `@FunctionalInterface`)

Assuming the implementing classes were respecting the contract (i.e., returning an `Assert` instance), the only breaking change is the method name changing from `createAssert` to `get`, which should not be a big deal. For example, the change on [`ObjectAssertFactory`](https://github.com/joel-costigliola/assertj-core/pull/1498/files/660461e1aa04f61e974554c8ba54ceb09ef50be8#diff-d14732ccd64a699c94b14a7b26ae9976) is just the method name.

Reply:
>IMHO without those changes the semantic of this interface was wrong, there was nothing ensuring that the produced object was an instance of `Assert`

On the contrary, different libraries might choose not to implement `org.assertj.core.api.Assert` and there is no need for an `AssertFactory` to enforce that. `AssertProvider` is something similar and doesn't enforce `Assert` as well.

> the only breaking change is the method name changing from `createAssert` to `get`, which should not be a big deal

Maybe it is not a big deal in an end user application project. However, there might be other libraries out there that are building on top of AssertJ. This would mean that if a user is using AssertJ and some other library the user would not be able to only upgrade AssertJ, but will need to wait for the other library to update to the latest version of AssertJ. Therefore, if I put on the hat as another library developer this is quite a big binary incompatible change.

P.S. I am not a member of the AssertJ team, only a contributor and a fellow open source developer. If the AssertJ team decides that this change is acceptable I am fine with it. I only wanted to voice my opinion

Reply:
No worries and I definitely appreciate your feedback :-) your point is absolutely reasonable. I challenged this because I introduced another breaking change in #1469 and it was considered fine.

Reply:
Just checked that one as well, that is an end user breaking change (it doesn't really affect other libraries building on top of AssertJ).

I don't know the scope of using `AssertFactory` though. Maybe it is like you are saying as well.

Reply:
I haven't had the time to look into this PR into details but I would rather avoid breaking changes unless it is not possible otherwise.

@scordio is it possible to keep `AssertFactory` as it is for that feature ? If it is not possible I'd rather prefer another interface to be defined independently of  `AssertFactory`.

Reply:
@joel-costigliola it is possible, but before doing it I'd like to stress this point a bit more.

There are two main changes here:
1. Restricting the base output type to `Assert` instead of `Object`
2. Inheriting from `Function`

About point 1.:
* PROS: right semantic (why an `AssertFactory` should return something that is not an `Assert` at all?)
* CONS: breaking change only if an implementing class was returning a type which was not a child of `Assert` (which didn't happen with any assertj internal classes, so internally the semantic was already implicitly right)

About point 2.:
* PROS: avoiding repeating what `Function` already provides and gaining [`Function#andThen`](https://docs.oracle.com/javase/8/docs/api/java/util/function/Function.html#andThen-java.util.function.Function-), [`Function#compose`](https://docs.oracle.com/javase/8/docs/api/java/util/function/Function.html#compose-java.util.function.Function-) and [`Function#identity`](https://docs.oracle.com/javase/8/docs/api/java/util/function/Function.html#identity--)
* CONS: bigger breaking change compared to 1., because the functional interface method name changes from `createAssert` to `get`. Who would be affected by this change is any class which implemented the functional interface explicitely instead of using lambdas or method reference. Who is not affected is any existing lambda and method reference already in place. Since Assertj 3 is Java 8 based, I see a quite reduced risk here.

If you are still not convinced, I'm totally fine to revert 1., 2. or both!

Reply:
@scordio I'm convinced for point 1 but not for point 2, I prefer keeping the `createAssert` method (the name is more expressive of the assertions domain) and I don't see a huge benefit for the inherited methods of `Function`.

Reply:
@joel-costigliola Done! You may want to do the same for `AssertProvider` (not needed for the feature in this PR).